enhancements/installer/azurestack.md

```diff
@@ -0,0 +1,160 @@
+
+## Motivation
+
+Currently, Azure Stack Hub is not supported but is a significant on-prem platform.
+
+### Goals
+
+1. Installer creates infrastructure and installs OpenShift on Azure Stack Hub.
+2. All standard support that entails, such as destroy.
+
+### Non-Goals
+
+1. ASH will not "inherit" established Azure
+features such as using existing networks or resource groups.
+
+## Proposal
+
+
+### User Stories
+
+#### Story 1
+As a cluster admin, I want to be able to use `openshift-install` to create and destroy a cluster on
+Azure Stack Hub.
+
+#### Story 2
+As an OpenShift developer, I want Azure Stack Hub to be managed and released like other platforms and covered by CI tests.
+
+### Implementation Details/Notes/Constraints [optional]
+
+Primary ways in which Azure Stack differs from Azure:
+- requires a different rhcos image
+- limited instance metadata service (IMDS) for VMs
+- does not support private DNS zones
+- limited subset of Azure infrastructure available (ergo different Terraform provider)
+
+### Risks and Mitigations
+
+Adding ASH support is somewhere in between adding a new Azure environment and an entirely new platform.
+
+## Design Details
+
+### Infrastructure Global Configuration
+
+#### API
+ASH will need to be added to the known [Azure Cloud Environments in the API](https://github.com/openshift/api/blob/master/config/v1/types_infrastructure.go#L339-L355):
+
+```go
+// AzureStackCloud is the Azure cloud environment used on-premises.
+AzureStackCloud AzureCloudEnvironment = "AzureStackCloud"
+```
+
+#### Cluster Config Operator
+Cluster config operator will [need to be updated](https://github.com/openshift/cluster-config-operator/blob/master/pkg/operator/kube_cloud_config/azure.go#L23) and the new API vendored in.
+
+#### API endpoints
+ASH endpoints are user-provided and, therefore, the Azure SDK treats ASH endpoints differently than the already-known endpoints of other Azure environments. When the cloud environment is set to
+`AZURESTACKCLOUD` the SDK expects the environment variable `AZURE_ENVIRONMENT_FILEPATH` to point to a [json configuration file](https://kubernetes-sigs.github.io/cloud-provider-azure/install/configs/#azure-stack-configuration), which is [typically located at `/etc/kuberentes/azurestackcloud.json`](https://github.com/kubernetes-sigs/cloud-provider-azure/issues/151).
+
+The Installer will need to construct this json configuration file from user input and include the file as part of the cloud-provider-config configmap. The file will also need to be present on nodes for the kubelet. `AZURE_ENVIRONMENT_FILEPATH` will need to be set programmatically for ASH on the kubelet, presumably through the kubelet's systemd unit [(see open questions)](#open-questions).
+
+Operators will set the `AZURE_ENVIRONMENT_FILEPATH` and mount the endpoints JSON file from the cloud-provider-config configmap. All operators using the Azure SDK will need to do this.
```

Comment:
Is there a way for operators to configure the Azure client directly with the custom endpoints rather than relying on a json file? It would be cumbersome for an operator with a static manifest to do this with mounts, given that the mounted ConfigMap must be copied over to the namespace. The mounting option is feasible for controllers that the operator starts but not easily for operators directly.

Reply:
Follow up on this. You can definitely set the endpoint programmatically when creating the client.
For example, see the following.
https://github.com/staebler/cluster-ingress-operator/blob/3bb8ed560892ae648806e6133d18803da9ec341b/pkg/dns/azure/client/client.go#L88

Reply:
@staebler I have updated this and I think these revisions make more sense.